File: workers/node/src/runner.ts
import type { Client } from "./client.js";
import type { HandlerResult } from "./result.js";
import type { ExternalTask } from "./types.js";

export type HandlerFn = (task: ExternalTask) => Promise<HandlerResult>;

export interface HandlerDefinition {
  topic: string;
  handle: HandlerFn;
}

/**
 * Builder form: `defineHandler({ topic, handle })`. The returned object is
 * also a HandlerDefinition you can pass to `runner.register(def)`.
 */
export function defineHandler(def: HandlerDefinition): HandlerDefinition {
  return def;
}

export interface RunnerConfig {
  workerId: string;
  maxJobs?: number;
  lockDurationSecs?: number;
  pollIntervalMs?: number;
}

export class Runner {
  private readonly handlers = new Map<string, HandlerFn>();
  private readonly cfg: Required<RunnerConfig>;
  private stopped = false;

  constructor(private readonly client: Client, config: RunnerConfig) {
    this.cfg = {
      workerId: config.workerId,
      maxJobs: config.maxJobs ?? 10,
      lockDurationSecs: config.lockDurationSecs ?? 30,
      pollIntervalMs: config.pollIntervalMs ?? 1_000,
    };
  }

  /** Register a handler. Accepts either `(topic, fn)` or a HandlerDefinition. */
  register(topicOrDef: string | HandlerDefinition, fn?: HandlerFn): void {
    if (typeof topicOrDef === "string") {
      if (!fn) throw new Error("register(topic, fn) requires fn");
      this.handlers.set(topicOrDef, fn);
    } else {
      this.handlers.set(topicOrDef.topic, topicOrDef.handle);
    }
  }

  stop(): void {
    this.stopped = true;
  }

  async run(): Promise<void> {
    if (this.handlers.size === 0) throw new Error("no handlers registered");
    while (!this.stopped) {
      const didWork = await this.tick();
      if (!didWork) {
        await sleep(this.cfg.pollIntervalMs);
      }
    }
  }

  /** Internal: one fetch-handle-report cycle across every registered topic. Exposed for tests. */
  async tick(): Promise<boolean> {
    let didWork = false;
    for (const topic of this.handlers.keys()) {
      let tasks: ExternalTask[];
      try {
        tasks = await this.client.fetchAndLock(
          this.cfg.workerId,
          topic,
          this.cfg.maxJobs,
          this.cfg.lockDurationSecs,
        );
      } catch (err) {
        console.error(`fetch-and-lock failed for ${topic}:`, err);
        continue;
      }
      for (const task of tasks) {
        await this.dispatch(task);
      }
      if (tasks.length > 0) didWork = true;
    }
    return didWork;
  }

  private async dispatch(task: ExternalTask): Promise<void> {
    const fn = this.handlers.get(task.topic ?? "");
    if (!fn) {
      console.warn(`no handler for topic ${task.topic} (task ${task.id})`);
      return;
    }
    let result: HandlerResult;
    try {
      result = await fn(task);
    } catch (err) {
      try {
        await this.client.failure(task.id, this.cfg.workerId, errMessage(err));
      } catch (ferr) {
        console.error(`failure call failed for task ${task.id}:`, ferr);
      }
      return;
    }
    try {
      if (result.kind === "bpmn-error") {
        await this.client.bpmnError(
          task.id,
          this.cfg.workerId,
          result.code,
          result.message,
          result.variables,
        );
      } else {
        await this.client.complete(task.id, this.cfg.workerId, result.variables);
      }
    } catch (err) {
      console.error(`report-back call failed for task ${task.id}:`, err);
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((res) => setTimeout(res, ms));
}

function errMessage(e: unknown): string {
  if (e instanceof Error) return e.message;
  return String(e);
}

File: workers/java/src/main/java/io/conduit/worker/Runner.java
package io.conduit.worker;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Fetch-handle-report loop. Register handlers, then call {@link #run}. */
public final class Runner {

  /** Tunes the polling loop. */
  public static final class Config {
    public final String workerId;
    public int maxJobs = 10;
    public int lockDurationSecs = 30;
    public Duration pollInterval = Duration.ofSeconds(1);

    public Config(String workerId) {
      this.workerId = workerId;
    }
  }

  private static final Logger LOG = Logger.getLogger(Runner.class.getName());

  private final Client client;
  private final Config cfg;
  private final ConcurrentHashMap<String, Handler> handlers = new ConcurrentHashMap<>();
  private final AtomicBoolean stopped = new AtomicBoolean(false);

  public Runner(Client client, Config cfg) {
    this.client = client;
    this.cfg = cfg;
  }

  /** Direct registration: bind {@code handler} to {@code topic}. */
  public Runner register(String topic, Handler handler) {
    handlers.put(topic, handler);
    return this;
  }

  /**
   * Annotation-driven registration: scan each instance for {@link TaskHandler}
   * and register it under the annotation's topic. Each instance must implement
   * {@link Handler}.
   */
  public Runner discover(Object... taskHandlers) {
    for (Object o : taskHandlers) {
      TaskHandler ann = o.getClass().getAnnotation(TaskHandler.class);
      if (ann == null) {
        throw new IllegalArgumentException(
            o.getClass().getName() + " is not annotated with @TaskHandler");
      }
      if (!(o instanceof Handler h)) {
        throw new IllegalArgumentException(
            o.getClass().getName() + " is annotated with @TaskHandler but does not implement Handler");
      }
      register(ann.topic(), h);
    }
    return this;
  }

  public void stop() {
    stopped.set(true);
  }

  public void run() throws InterruptedException {
    if (handlers.isEmpty()) throw new IllegalStateException("no handlers registered");
    while (!stopped.get()) {
      boolean didWork = tick();
      if (!didWork) {
        Thread.sleep(cfg.pollInterval.toMillis());
      }
    }
  }

  /** One fetch-handle-report cycle across every registered topic. Public for tests. */
  public boolean tick() throws InterruptedException {
    boolean didWork = false;
    for (String topic : handlers.keySet()) {
      List<ExternalTask> tasks;
      try {
        tasks = client.fetchAndLock(cfg.workerId, topic, cfg.maxJobs, cfg.lockDurationSecs);
      } catch (Exception e) {
        LOG.log(Level.WARNING, "fetch-and-lock failed for topic " + topic, e);
        continue;
      }
      for (ExternalTask t : tasks) {
        dispatch(t);
      }
      if (!tasks.isEmpty()) didWork = true;
    }
    return didWork;
  }

  private void dispatch(ExternalTask task) {
    Handler fn = handlers.get(task.topic());
    if (fn == null) {
      LOG.warning("no handler for topic " + task.topic() + " (task " + task.id() + ")");
      return;
    }
    HandlerResult result;
    try {
      result = fn.handle(task);
    } catch (Exception ex) {
      try {
        client.failure(task.id(), cfg.workerId, ex.getMessage() == null ? ex.toString() : ex.getMessage());
      } catch (Exception ferr) {
        LOG.log(Level.SEVERE, "failure call failed for task " + task.id(), ferr);
      }
      return;
    }
    try {
      switch (result) {
        case HandlerResult.BpmnError be ->
            client.bpmnError(task.id(), cfg.workerId, be.code(), be.message(), be.variables());
        case HandlerResult.Complete c ->
            client.complete(task.id(), cfg.workerId, c.variables());
      }
    } catch (Exception ex) {
      LOG.log(Level.SEVERE, "report-back call failed for task " + task.id(), ex);
    }
  }
}
